### PR TITLE
Addresses missing test coverage

### PR DESCRIPTION
**Pull request by HDembinski**
We want moar tests.

Addresses #242, #241, #239.

### CODE REVIEW

**Comment by HDembinski** [comment]
@cdeil In case you want to review this, the diff may look very complicated, but the new code is much simpler and more clear than the old one.

File: iminuit/tests/test_iminuit.py [hunk 1]
```diff
@@ -246,63 +271,31 @@ def f(x, y):
...
-    assert_allclose(val['z'], 7)
-    assert_allclose(m.fval, 10. + 2.5)
-    free_param = m.list_of_vary_param()
-    fix_param = m.list_of_fixed_param()
-    assert 'x' in free_param
-    assert 'x' not in fix_param
-    assert 'y' in fix_param
-    assert 'y' not in free_param
-    assert 'z' not in fix_param
+    for grad in (None, func3_grad):
```
Comment (cdeil) [inline iminuit/tests/test_iminuit.py]: Use https://docs.pytest.org/en/latest/parametrize.html here instead of for loop in the test?

Comment (HDembinski) [reply iminuit/tests/test_iminuit.py]: Good point, I just read about this yesterday. I love pytest, I must say. It is really pythonic.

**Comment by HDembinski** [comment]
I use `pytest.mark.parametrize` now to test various conditions, which is very elegant. Once the CI tests ran through, I will merge this.

**Pull request merged**